bench/richards/richards.dora

```diff
@@ -0,0 +1,347 @@
+// Dora implementation of Richards benchmark. 
+// Originally implemented in BCPL (http://www.cl.cam.ac.uk/~mr10/Bench.html)
+// Used in Octane 2.0 (https://developers.google.com/octane/benchmark)
+
+fun main() {
+    let bench = Benchmark();
+
+    let iterations = 8200;
+    var i = 0;
+    while i < iterations {
+        bench.runRichards();
+        i = i + 1;
+    }
+}
+
+var COUNT: Int32 = 1000;
+var EXPECTED_QUEUE_COUNT: Int32 = 2322;
+var EXPECTED_HOLD_COUNT: Int32 = 928;
+
+class Benchmark() {
+    fun runRichards() {
+        let scheduler = Scheduler();
+        scheduler.addIdleTask(ID_IDLE, 0, none[Packet](), COUNT);
+
+        var queue = Packet::new(none[Packet](), ID_WORKER, KIND_WORK);
+        queue = Packet::new(some[Packet](queue), ID_WORKER, KIND_WORK);
+        scheduler.addWorkerTask(ID_WORKER, 1000, some[Packet](queue));        
+
+        queue = Packet::new(none[Packet](), ID_DEVICE_A, KIND_DEVICE);
+        queue = Packet::new(some[Packet](queue), ID_DEVICE_A, KIND_DEVICE);
+        queue = Packet::new(some[Packet](queue), ID_DEVICE_A, KIND_DEVICE);
+        scheduler.addHandlerTask(ID_HANDLER_A, 2000, some[Packet](queue));
+
+        queue = Packet::new(none[Packet](), ID_DEVICE_B, KIND_DEVICE);
+        queue = Packet::new(some[Packet](queue), ID_DEVICE_B, KIND_DEVICE);
+        queue = Packet::new(some[Packet](queue), ID_DEVICE_B, KIND_DEVICE);
+        scheduler.addHandlerTask(ID_HANDLER_B, 3000, some[Packet](queue));
+
+        scheduler.addDeviceTask(ID_DEVICE_A, 4000, none[Packet]());
+
+        scheduler.addDeviceTask(ID_DEVICE_B, 5000, none[Packet]());
+
+        scheduler.schedule();
+
+        if scheduler.queueCount != EXPECTED_QUEUE_COUNT || scheduler.holdCount != EXPECTED_HOLD_COUNT {
+            println("Error during execution: queueCount = ${scheduler.queueCount}, holdCount = ${scheduler.holdCount}.");
+            exit(1);
+        }
+    }
+}
+
+var ID_IDLE: Int32 = 0;
+var ID_WORKER: Int32 = 1;
+var ID_HANDLER_A: Int32 = 2;
+var ID_HANDLER_B: Int32 = 3;
+var ID_DEVICE_A: Int32 = 4;
+var ID_DEVICE_B: Int32 = 5;
+var NUMBER_OF_IDS: Int32 = 6;
+
+var KIND_DEVICE: Int32 = 0;
+var KIND_WORK: Int32 = 1;
+
+class Scheduler() {
+    var queueCount: Int32 = 0;
+    var holdCount: Int32 = 0;
+    var blocks: Array[Option[TaskControlBlock]] = arrayFill[Option[TaskControlBlock]](NUMBER_OF_IDS.toInt64(), none[TaskControlBlock]());
+    var currentTcb: Option[TaskControlBlock] = none[TaskControlBlock]();
```

Comment:
Option is still quite expensive in Dora relative to using `nil`, so I don't expect the Dora version to be competitive with that of JS. But we can keep the code like this. IIRC for binarytrees I measured a slowdown of factor 5, it's probably less here though.

Reply:
I recall you mentioning the expensive unwrapping of `Option`, but decided to use it anyway, as `nil` could possibly be removed in the long run.